stream/stream.go

```diff
@@ -4,229 +4,230 @@ package stream
 
 import (
 	"context"
-	"encoding/binary"
 	"encoding/json"
 	"errors"
 	"fmt"
+	"sync"
 	"time"
 
 	"github.com/lthibault/log"
 	"github.com/prometheus/client_golang/prometheus"
 
+	"github.com/blocknative/dreamboat/stream/transport"
 	"github.com/blocknative/dreamboat/structs"
 	"github.com/blocknative/dreamboat/structs/forks/bellatrix"
 	"github.com/blocknative/dreamboat/structs/forks/capella"
 )
 
-var (
+const (
 	CacheTopic         = "/block/cache"
 	BidTopic           = "/block/bid"
 	SlotDeliveredTopic = "/slot/delivered"
 )
 
-type Pubsub interface {
-	Publish(context.Context, string, []byte) error
-	Subscribe(context.Context, string) chan []byte
-}
-
-type StreamConfig struct {
-	Logger          log.Logger
-	ID              string
-	PubsubTopic     string // pubsub topic name for block submissions
-	StreamQueueSize int
+type PubSub interface {
+	Publish(context.Context, transport.Message) error
+	Subscribe(context.Context) transport.Subscription
 }
 
 type State interface {
 	ForkVersion(epoch structs.Slot) structs.ForkVersion
 	HeadSlot() structs.Slot
 }
 
+type Metrics interface {
+	Register(prometheus.Collector) error
+}
+
 type Client struct {
-	Pubsub Pubsub
+	once sync.Once
 
-	builderBidIn     chan []byte
-	builderBidOut    chan structs.BuilderBidExtended
-	cacheIn          chan []byte
-	cacheOut         chan structs.BlockAndTraceExtended
-	slotDeliveredIn  chan []byte
-	slotDeliveredOut chan uint64
+	Logger                log.Logger
+	State                 State
+	Bids, Cache           PubSub
+	QueueSize, NumWorkers int
 
-	Config StreamConfig
-	Logger log.Logger
+	Metrics Metrics
+	m       streamMetrics
 
-	m StreamMetrics
+	ctx    context.Context
+	cancel context.CancelFunc
 
-	//slotDelivered chan structs.Slot
+	builderBidOut chan structs.BuilderBidExtended
+	cacheOut      chan structs.BlockAndTraceExtended
 
 	st State
 }
 
-func NewClient(ps Pubsub, st State, cfg StreamConfig) *Client {
-	s := Client{
-		Pubsub: ps,
-		st:     st,
-
-		builderBidIn:     make(chan []byte, cfg.StreamQueueSize),
-		builderBidOut:    make(chan structs.BuilderBidExtended, cfg.StreamQueueSize),
-		cacheIn:          make(chan []byte, cfg.StreamQueueSize),
-		cacheOut:         make(chan structs.BlockAndTraceExtended, cfg.StreamQueueSize),
-		slotDeliveredIn:  make(chan []byte, cfg.StreamQueueSize),
-		slotDeliveredOut: make(chan uint64, cfg.StreamQueueSize),
+func (c *Client) init() {
+	c.once.Do(func() {
```

Comment:
I'm not really a fan of once. But regardless - you should be able to instantialize this as much as you like as it doesn't hold global context - I don't see any reason why you would need one here. 
And being - in main.go why would I like to do it on every function call this structure have (BlockCache, BuilderBid,  PublishBuilderBid)?  
New was just meant to safely initialize what ever client may need in future, so that writing any additional function I should not pay attention to some initializer someone may like me to call.

Reply:
I'm quite attached to this change because a `New*` function is going to be quite clunky due to the large number of arguments.

> you should be able to instantialize this as much as you like as it doesn't hold global context 

I'm not sure I understand.  The lazy initialization is needed to set up the goroutines and allocate the channels.

>New was just meant to safely initialize what ever client may need in future, so that writing any additional function I should not pay attention to some initializer someone may like me to call.

Functionally, this is exactly equivalent to `New*`.  You just instantiate a struct literal instead of making a function call, but there's no additional step from the point of view of the public API.

Reply:
You call init() in every public function `BlockCache`, `BuilderBid`, `PublishBuilderBid` -  instead of just initializing it once and using from the outside  - not need for knowing/checking anything as you know everything is initialized. And New here exists for the very reason of making all the channels and everything that may possibly effect in nil pointer exception of or deadlock. It also gives you a way not to use it if you, for some reason prefer to bypass that and do it yourself.